File: 2.add-two-numbers.java

import java.util.List;


/*
 * @lc app=leetcode id=2 lang=java
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode hangToa1, ListNode hangToa2) {
        ListNode dummyHead = new ListNode(0); // Tạo nút giả để dễ quản lý
        ListNode current = dummyHead; // Biến để duyệt qua danh sách mới   
        int carry = 0; // Biến để lưu trữ giá trị nhớ khi cộng
        while (hangToa1 != null || hangToa2 != null || carry >
                0) { // Tiếp tục khi còn nút trong danh sách hoặc còn giá trị nhớ
            int sum = carry; // Bắt đầu với giá trị nhớ
            if (hangToa1 != null) { // Nếu còn nút trong danh sách đầu tiên
                sum += hangToa1.val; // Cộng giá trị của nút vào tổng
                hangToa1 = hangToa1.next; // Di chuyển đến nút tiếp theo
            }
            if (hangToa2 != null) { // Nếu còn nút trong danh sách thứ hai
                sum += hangToa2.val; // Cộng giá trị của nút vào tổng
                hangToa2 = hangToa2.next; // Di chuyển đến nút tiếp theo
            }
            carry = sum / 10; // Tính giá trị nhớ cho lần cộng tiếp theo
            current.next = new ListNode(sum % 10); // Tạo nút mới với giá trị là phần dư
            current = current.next; // Di chuyển đến nút mới
        }
        return dummyHead.next; // Trả về danh sách kết quả (bỏ qua nút giả)
    }
}

// @lc code=end

File: 2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python
#
# [2] Add Two Numbers
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        dummy = ListNode(0) # Tạo một nút giả để bắt đầu danh sách kết quả
        current = dummy # Biến current sẽ dùng để xây dựng danh sách kết quả
        carry = 0 # Biến carry để lưu trữ giá trị chuyển tiếp khi tổng vượt quá 10

        while l1 or l2 or carry: # Duyệt qua cả hai danh sách và kiểm tra xem có giá trị carry không
            val1 = l1.val if l1 else 0 # Lấy giá trị của nút l1, nếu không có thì dùng 0
            val2 = l2.val if l2 else 0 # Lấy giá trị của nút l2, nếu không có thì dùng 0
            total = val1 + val2 + carry # Tính tổng của hai giá trị và giá trị carry
            carry = total // 10 # Cập nhật giá trị carry
            current.next = ListNode(total % 10) # Tạo nút mới với giá trị là tổng modulo 10
            current = current.next # Di chuyển current đến nút mới
            if l1: # Di chuyển l1 đến nút tiếp theo nếu có
                l1 = l1.next 
            if l2: # Di chuyển l2 đến nút tiếp theo nếu có
                l2 = l2.next

        return dummy.next # Trả về danh sách kết quả bắt đầu từ nút tiếp theo của nút giả